### PR TITLE
[HIP-0026] Plugin runtime interface

### PR DESCRIPTION
**Pull request by scottrigby**
Adds plugin Runtime abstraction to encapsulate subprocess code and enable future runtimes.

Part of implementing hip-0026. For full context, see:
- https://github.com/helm/helm/issues/31144

Depends on:
- #31142
Please review that PR first. After that PR is merged, will change this PR base branch to `main` and move out of draft.

X This PR does NOT contain user facing changes

- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: internal/plugin/runtime_subprocess_getter.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		msg.Options.KeyFile,
+		msg.Options.CAFile,
+		msg.Href)
+
+	// TODO should we append to input.Env too?
+	//env := append(os.Environ(), input.Env...)
+	//env = append(env,
+	//	fmt.Sprintf("HELM_PLUGIN_USERNAME=%s", msg.Options.Username),
+	//	fmt.Sprintf("HELM_PLUGIN_PASSWORD=%s", msg.Options.Password),
+	//	fmt.Sprintf("HELM_PLUGIN_PASS_CREDENTIALS_ALL=%t", msg.Options.PassCredentialsAll))
```
Comment (jessesimpson36) [inline internal/plugin/runtime_subprocess_getter.go]: seems like this comment doesn't need to be here since it's uncommented 2 lines down.

File: internal/plugin/loader.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+// filterFunc is a function that filters plugins
+type filterFunc func(Plugin) bool
+
+// FindPlugins returns a list of plugins that match the descriptor
+func FindPlugins(pluginsDirs []string, descriptor Descriptor) ([]Plugin, error) {
+	return findPlugins(pluginsDirs, LoadAll, makeDescriptorFilter(descriptor))
+}
+
+// findPlugins is the internal implementation that uses the find and filter functions
+func findPlugins(pluginsDirs []string, findFunc findFunc, filterFunc filterFunc) ([]Plugin, error) {
```
Comment (jessesimpson36) [inline internal/plugin/loader.go]: I'd suggest the name findFunc not be of type findFunc, otherwise you're shadowing over the type declaration with a variable name, and I don't think you could create another of type findFunc in this body.

File: internal/plugin/metadata.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	SourceURL string
+
+	// Config contains the type-specific configuration for this plugin
+	Config Config
+
+	// RuntimeConfig contains the runtime-specific configuration
+	RuntimeConfig RuntimeConfig
+}
+
+func (m Metadata) Validate() error {
```
Comment (jessesimpson36) [inline internal/plugin/metadata.go:51]: This Validate function returns at the first error it sees. I could imagine a more helpful Validate function returning all the validation errors all at once, sortof how the way linters in editors work. Perhaps this should return a list of errors rather than just an error? or could wrap a list of errors.

**Review by jessesimpson36 — APPROVED** [review]
I've left comments, but I still approve. 

Overall, I've been testing this out manually, and everything seems to work fine. The unit tests appear to cover the happy-paths well, and could easily be copy/pasted later to test any missing error edge-cases, or used as helper documentation for a user to make your own plugin (that's how I've been making one of my test plugins). 

I reviewed the HIP once more, and these changes appear to abide by what was declared in the HIP, no surprising changes.

**Comment by jessesimpson36** [comment]
Test downloader plugin I wrote from this PR
https://github.com/jessesimpson36/helm-serial-getter

File: internal/plugin/runtime_subprocess_getter.go (hunk 1 above)
Comment (scottrigby) [reply internal/plugin/runtime_subprocess_getter.go]: I agree. Will trim it. 

BTW, @gjenkins8 found a novel way to address this TODO in a PR on my fork. We're keeping the upstream PRs as tidy as possible since several of them have sequential depednencies. If you're curious, it's that fix is here: https://github.com/scottrigby/helm/pull/24 We'll likely wait until after the pre-release freeze to make an upstream PR for those changes.

File: internal/plugin/loader.go (hunk 2 above)
Comment (scottrigby) [reply internal/plugin/loader.go]: good point. will fix this variable shadowing issue.

File: internal/plugin/metadata.go (hunk 3 above)
Comment (scottrigby) [reply internal/plugin/metadata.go:51]: good idea. we use errors.Join() elsewhere too (eg, chart validation). done

Comment (joejulian) [reply internal/plugin/metadata.go:51]: Did we ever do this?
https://github.com/helm/helm/pull/12298

File: internal/plugin/loader.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+func newPrototypePluginManager() *prototypePluginManager {
+	return &prototypePluginManager{
+		runtimes: map[string]Runtime{
+			"subprocess": &RuntimeSubprocess{},
+		},
+	}
+}
+
+func (pm *prototypePluginManager) RegisterRuntime(runtimeName string, runtime Runtime) {
+	//if _, exists := pm.runtimes[runtimeName]; exists {
```
Comment (joejulian) [inline internal/plugin/loader.go]:
```suggestion
```

File: internal/plugin/metadata_test.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		{true, mockSubprocessCLIPlugin(t, "abcdefghijklmnopqrstuvwxyz0123456789_-ABC")},
+		{true, mockSubprocessCLIPlugin(t, "foo-bar-FOO-BAR_1234")},
+		{false, mockSubprocessCLIPlugin(t, "foo -bar")},
+		{false, mockSubprocessCLIPlugin(t, "$foo -bar")}, // Test leading chars
+		{false, mockSubprocessCLIPlugin(t, "foo -bar ")}, // Test trailing chars
+		{false, mockSubprocessCLIPlugin(t, "foo\nbar")},  // Test newline
+		{true, mockNoCommand},     // Test no command metadata works
+		{true, mockLegacyCommand}, // Test legacy command metadata works
+		{false, mockWithCommand},  // Test platformCommand and command both set fails
+		{false, mockWithHooks},    // Test platformHooks and hooks both set fails
```
Comment (joejulian) [inline internal/plugin/metadata_test.go]: You've done the work of describing the test, why not make that string part of the test data so the result can say which test failed (or succeeded)?

**Review by joejulian — DISMISSED** [review]

File: internal/plugin/metadata.go (hunk 3 above)
Comment (scottrigby) [reply internal/plugin/metadata.go:51]: > Did we ever do this? #12298

Yes, earlier this year. I think your PR should have been merged 2 yrs ago, I think it just got missed. Added some history here: https://github.com/helm/helm/pull/12298#issuecomment-3212770756 🙂

File: internal/plugin/loader.go (hunk 2 above)
Comment (scottrigby) [reply internal/plugin/loader.go]: fixed

File: internal/plugin/metadata_test.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	} {
+		err := item.plug.Metadata().Validate()
+		if item.pass && err != nil {
+			t.Errorf("failed to validate case %d: %s", i, err)
+		} else if !item.pass && err == nil {
+			t.Errorf("expected case %d to fail", i)
+		}
+	}
+}
```
Comment (jessesimpson36) [inline internal/plugin/metadata_test.go]:
```suggestion
	for i, item := range []struct {
		pass      bool
		plug      Plugin
		errString string
	}{
		{true, mockSubprocessCLIPlugin(t, "abcdefghijklmnopqrstuvwxyz0123456789_-ABC"), ""},
		{true, mockSubprocessCLIPlugin(t, "foo-bar-FOO-BAR_1234"), ""},
		{false, mockSubprocessCLIPlugin(t, "foo -bar"), "invalid name"},
		{false, mockSubprocessCLIPlugin(t, "$foo -bar"), "invalid name"}, // Test leading chars
		{false, mockSubprocessCLIPlugin(t, "foo -bar "), "invalid name"}, // Test trailing chars
		{false, mockSubprocessCLIPlugin(t, "foo\nbar"), "invalid name"},  // Test newline
		{true, mockNoCommand, ""},     // Test no command metadata works
		{true, mockLegacyCommand, ""}, // Test legacy command metadata works
		{false, mockWithCommand, "runtime config validation failed: both platformCommand and command are set"}, // Test platformCommand and command both set fails
		{false, mockWithHooks, "runtime config validation failed: both platformHooks and hooks are set"},       // Test platformHooks and hooks both set fails
	} {
		err := item.plug.Metadata().Validate()
		if item.pass && err != nil {
			t.Errorf("failed to validate case %d: %s", i, err)
		} else if !item.pass && err == nil {
			t.Errorf("expected case %d to fail", i)
		}
		if !item.pass && err.Error() != item.errString {
			t.Errorf("index [%d]: expected the following error: %s, but got: %s", i, item.errString, err.Error())
		}
	}
}
```

File: internal/plugin/metadata_test.go (hunk 5 above)
Comment (scottrigby) [reply internal/plugin/metadata_test.go]: I committed @jessesimpson36 's code suggestion addressing this 1238a78300d15691e9b5122a7db22dfbc5173e96. Is this what you meant?

File: internal/plugin/loader.go (hunk 4 above)
Comment (scottrigby) [reply internal/plugin/loader.go]: committed

File: internal/plugin/metadata_test.go (hunk 6 above)
Comment (scottrigby) [reply internal/plugin/metadata_test.go]: committed

**Review by joejulian — APPROVED** [review]

**Comment by scottrigby** [comment]
Thanks y'all, merging so we can tee up the next PR in the list on #31144
